### ISSUE TITLE
x/vgo: sync direct imports of transitive dependencies

### ISSUE THREAD

**Issue by mwf** [golang/go#25969]
### What version of Go are you using (`go version`)?
go version go1.10.1 darwin/amd64 vgo:2018-02-20.1

Latest `vgo`:
```
commit f574d316627652354b92fbdf885a2b2f7ea7dac9 (HEAD -> master, origin/master, origin/HEAD)
Author: Bryan C. Mills <bcmills@google.com>
Date:   Fri Jun 15 16:41:09 2018 -0400

    cmd/go: skip vgo.Init for the version subcommand
```

### What did you do?

Please refer to https://github.com/mwf/goplay/tree/master/vgo/transitive

Consider such a flow:
1. Project is empty, you create `main.go` with the only dependency of `github.com/mwf/goplay/vgo/vtest/bar`
```go
package main

import (
	"github.com/mwf/goplay/vgo/vtest/bar"
)

func main() {
	println("bar", bar.Bar)
	println("foobar", bar.FooBar)
}
```

2. Create a go.mod `vgo mod -init` and sync it `vgo mod -sync`. We get the only one dependency `bar` of latest version and `foo` as a transitive, included in bar's `go.mod`

```
$ cat go.mod
module github.com/mwf/goplay/vgo/transitive

require github.com/mwf/goplay/vgo/vtest/bar v0.1.0
```
```
$ vgo list -m
MODULE                                VERSION
github.com/mwf/goplay/vgo/transitive  -
github.com/mwf/goplay/vgo/vtest/bar   v0.1.0
github.com/mwf/goplay/vgo/vtest/foo   v0.1.0
```

3. Some time passes, and you need to import the `github.com/mwf/goplay/vgo/vtest/foo` directly in your app.
main.go:
```go
package main

import (
	"github.com/mwf/goplay/vgo/vtest/bar"
	"github.com/mwf/goplay/vgo/vtest/foo"
)

func main() {
	println("bar", bar.Bar)
	println("foobar", bar.FooBar)
	println("foo", foo.Foo)
}
```

4. And you want to sync the dependencies with `vgo mod -sync`

### What did you expect to see?

I expect `vgo mod -sync` to update `go.mod` with new requirement `github.com/mwf/goplay/vgo/vtest/foo   v0.1.0`. Because it stopped being transitive and became a direct dependency.

### What did you see instead?

`vgo mod -sync` does nothing, keeping `github.com/mwf/goplay/vgo/vtest/foo` as transitive.

### Rationale

I think we should populate go.mod file with all **direct** dependencies found in the code. 
Otherwise there could be bad consequences for long-term project support.

For example, we remove `bar` usage from code in some future release, only `foo` is left. We would definitely want to drop bar from `go.mod` - and we delete it with `vgo mod -droprequire=github.com/mwf/goplay/vgo/vtest/bar` or manually. Running `vgo mod -sync` then resolves `foo` to the **latest** version, which is not we probably want.

There is a way to handle it now - instead of `droprequire` or manual go.mod modification run `vgo mod -sync` an it keeps the transitive version used previously. But better to support all the cases.

Summarising, it's better to have all **direct** dependencies **explicitly defined** in go.mod instead of keeping in mind some _correct_ way of dependency synchronization.

**Comment by bcmills** [comment]
> I think we should populate go.mod file with all *direct* dependencies found in the code.
> Otherwise there could be bad consequences for long-term project support.

Coincidentally, Russ and I were discussing this yesterday (in the context of [CL 119575](https://golang.org/cl/119575)) and came to pretty much the same conclusion. If we don't list all of the direct dependencies, then `mvs.UpgradeAll` can end up dropping dependencies that we have to add back in, and `go get -u` becomes much more complicated (and probably iterative).

On the other hand, if we include these implied requirements, we have to be careful to keep them accurate. If module `A` imports `C` and requires module `B@v1.1.0`, and `B@v1.1.0` requires `C@v1.1.0`, then module `A` must require `C@v1.1.0` or higher: `C@v1.0.0` would not satisfy the requirements of `B@v1.1.0`, so listing `C@v1.0.0` as a requirement would be misleading.

**Comment by mwf** [comment]
> On the other hand, if we include these implied requirements, we have to be careful to keep them accurate. If module A imports C and requires module B@v1.1.0, and B@v1.1.0 requires C@v1.1.0, then module A must require C@v1.1.0 or higher: C@v1.0.0 would not satisfy the requirements of B@v1.1.0, so listing C@v1.0.0 as a requirement would be misleading.

Agree, we should do it carefully. 
But actually it should be easy enough, because as I said `vgo mod -sync` already does what we need - it adds an already resolved transitive dependency (max of mins) if you drop calls to `B` from your code leaving `require B v1.1.0` in go.mod. - it will change to `require C v1.1.0`

So you should only translate this logic to the case when `B` is still used along with `C` :)

**Comment by mwf** [comment]
I'm sorry @rsc, but I see https://golang.org/cl/121304 is still not merged to master. 

Could we please reopen this issue till it's merged? Just want to track the progress here and check it when finally done. 
Thanks in advance :)

**Comment by bcmills** [comment]
@mwf, looks like it was merged in commit `552f8db34257fd0182aa02302bc29db1aeef7370`.

**Comment by mwf** [comment]
Yeap, I'm already using it, thanks 👍
